spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyManager.java

```diff
@@ -0,0 +1,122 @@
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin = minutes;
+    }
+
+    public ApiKey findApiKeyFromLocalApiKeyList(String apiKeyValue) {
+        log.debug("Trying to find the service API associated to the API key with the value {} from the local list.",
+                apiKeyValue);
+
+        if (apiKeyValue == null) {
+            log.info("The passed API key value is null and can therefore not be checked for validity. " +
+                    "It is therefore rejected.");
+            return null;
+        }
+        Optional<ApiKey> optionalApiKey = localApiKeyList.stream().filter(
+                apiKeyFromLocalList -> apiKeyFromLocalList.getKeyValue().equals(apiKeyValue)).findFirst();
+        return extractOptionalApiKeyAndLogPossibleErrors(optionalApiKey, apiKeyValue);
+    }
+
+    private ApiKey extractOptionalApiKeyAndLogPossibleErrors(Optional<ApiKey> optionalApiKey, String apiKeyValue) {
+        if (optionalApiKey.isPresent()) {
+            log.debug("The API key with the value {} is found in the local API key list.", apiKeyValue);
+            return optionalApiKey.get();
+        } else {
+            log.info("The API key with the value {} can not be found within the local API key list " +
+                    "and is therefore rejected.", apiKeyValue);
+            return null;
+        }
+    }
+
+    public boolean wasLatestUpdateOfLocalApiKeyListWithinDeadline(ApiKey apiKey) {
+        LocalDateTime deadline = latestLocalApiKeyListUpdate.plusMinutes(
+                numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin);
+        boolean wasLatestUpdateWithinDeadline = LocalDateTime.now().isBefore(deadline);
+
+        if (wasLatestUpdateWithinDeadline)
+            log.info("The API key with the value {} is valid considering the local API key list and is " +
+                    "therefore accepted.", apiKey.getKeyValue());
+        else
+            log.info("The CoatRack admin server was not reachable for longer than {} minutes. Since this " +
+                    "predefined threshold was exceeded, this and all subsequent service API requests are " +
+                    "rejected until a connection to CoatRack admin could be re-established.",
+                    numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin);
+
+        return wasLatestUpdateWithinDeadline;
+    }
+
+    @Async
+    @PostConstruct
+    @Scheduled(fixedRateString = "${local-api-key-list-update-interval-in-millis}")
+    public void updateLocalApiKeyList() {
+        List<ApiKey> apiKeys;
+        try {
+            apiKeys = apiKeyFetcher.requestLatestApiKeyListFromAdmin();
+        } catch (ConnectException e) {
+            log.info("Trying to update the local API key list, the connection to CoatRack admin failed.");
+            log.debug("Following error occurred: " + e);
+            return;
+        }
+        localApiKeyList = apiKeys;
```

Comment:
This and the following line should be moved into the `try` block, because they should only be executed if requesting the list from admin was successful.

Reply:
That is why I added the return statement within the catch-block. try-blocks should only contain risky code, which is why i added the non-risky code below.